scripts/internal/prune_size_model.py

```diff
@@ -0,0 +1,310 @@
+
+        next_larger_num_xgrams = cur_target_num_xgrams
+        while left <= right - tolerance:
+            next_threshold = (left + right) / 2
+            modeled_next_num_xgrams = self.GetModeledNextNumXgrams(next_threshold)
+
+            if modeled_next_num_xgrams < cur_target_num_xgrams:
+                right = next_threshold
+            elif modeled_next_num_xgrams > cur_target_num_xgrams:
+                next_larger_num_xgrams = modeled_next_num_xgrams
+                left = next_threshold
+            else: # modeled_next_num_xgrams == cur_target_num_xgrams, this will probably not happan
+                next_larger_num_xgrams = modeled_next_num_xgrams
+                break
+        if left > right - tolerance:
+            # the while loop is not breaked by the else clause,
+            # so we make sure the modeled_next_num_xgrams >= cur_target_num_xgrams
+            next_threshold = left
+
+        if math.fabs(cur_target_num_xgrams - next_larger_num_xgrams) > 0.2 * cur_target_num_xgrams:
+            LogMessage("Warning: final modeled_next_num_xgrams[{0}] is far "
+                       "from intermediate_target_num_xgrams{1}".format(
+                           next_larger_num_xgrams, cur_target_num_xgrams))
+
+        return (next_threshold, next_larger_num_xgrams)
+
+    def GetModeledNextNumXgrams(self, next_threshold):
+        """
+        This function is a model of how we think the next num-xgrams
+        will vary with the chosen next threshold.
+        We predict the next_num_xgrams based on num_xgrams changed of
+        previous iteration and changes of threshold.
+        """
+
+        prev_num_xgrams = self.GetPrevNumXgrams()
+        cur_threshold = self.GetCurThreshold()
+        cur_num_xgrams = self.GetCurNumXgrams()
+
+        # First predict the num-xgrams we think we'll get if we prune again
+        # with the same threshold 'cur_threshold'. The basic heuristic is
+        # that the num-xgrams will decrease by a factor no greater than
+        # 1.5, and no greater than the cube root of the factor by which
+        # the num-xgrams decreased on the previous iteration.
+        assert prev_num_xgrams >= cur_num_xgrams
+        prev_change_factor = (float(cur_num_xgrams) / prev_num_xgrams) ** (1.0/3.0)
+        baseline_decrease_factor = 1.0 / 1.5
+        # choose the factor from these two, that's closest to one.
+        predicted_decrease_factor = max(prev_change_factor, baseline_decrease_factor)
+
+        # the following gives us the predicted num-xgrams if we were
+        # to prune again with the same threshold 'cur_threshold'.
+        predicted_num_xgrams_if_repeat = predicted_decrease_factor * cur_num_xgrams
+        assert next_threshold >= cur_threshold
+
+        predicted_extra_factor = (next_threshold / cur_threshold) ** (self.xgrams_change_power)
+        return predicted_num_xgrams_if_repeat * predicted_extra_factor
+
+    def AdjustModelForOvershoot(self):
```

Comment:
OK, I will stash the code of that part of modification. I hope the problems won't happen if i corrently implement the backtracking.